adapters/adapterstest/test_json.go

```diff
@@ -107,23 +109,67 @@ func runSpec(t *testing.T, filename string, spec *testSpec, bidder adapters.Bidd
 	} else if isVideoTest {
 		reqInfo.PbsEntryPoint = "video"
 	}
-	actualReqs, errs := bidder.MakeRequests(&spec.BidRequest, &reqInfo)
-	diffErrorLists(t, fmt.Sprintf("%s: MakeRequests", filename), errs, spec.MakeRequestErrors)
-	diffHttpRequestLists(t, filename, actualReqs, spec.HttpCalls)
 
-	bidResponses := make([]*adapters.BidderResponse, 0)
+	actualReqs := testMakeRequestsImpl(t, filename, spec, bidder, &reqInfo)
+
+	testMakeBidsImpl(t, filename, spec, bidder, actualReqs)
+}
+
+func testMakeRequestsImpl(t *testing.T, filename string, spec *testSpec, bidder adapters.Bidder, reqInfo *adapters.ExtraRequestInfo) []*adapters.RequestData {
+	t.Helper()
+
+	// Save original bidRequest values to assert no data races occur inside MakeRequests latter
+	deepBidReqCopy := openrtb.BidRequest{}
+	copier.Copy(&deepBidReqCopy, &spec.BidRequest)
+	shallowBidReqCopy := spec.BidRequest
+
+	// Save original []Imp elements to assert no data races occur inside MakeRequests latter
+	deepImpCopies := make([]openrtb.Imp, len(spec.BidRequest.Imp))
+	shallowImpCopies := make([]openrtb.Imp, len(spec.BidRequest.Imp))
+	for i := 0; i < len(spec.BidRequest.Imp); i++ {
+		deepImpCopy := openrtb.Imp{}
+		copier.Copy(&deepImpCopy, &spec.BidRequest.Imp[i])
+		deepImpCopies = append(deepImpCopies, deepImpCopy)
+
+		shallowImpCopy := spec.BidRequest.Imp[i]
+		shallowImpCopies = append(shallowImpCopies, shallowImpCopy)
+	}
+
+	// Run MakeRequests
+	actualReqs, errs := bidder.MakeRequests(&spec.BidRequest, reqInfo)
+
+	// Compare MakeRequests actual output versus expected values found in JSON file
+	assertErrorList(t, fmt.Sprintf("%s: MakeRequests", filename), errs, spec.MakeRequestErrors)
+	assertMakeRequestsOutput(t, filename, actualReqs, spec.HttpCalls)
+
+	// Assert no data races occur using original bidRequest copies of references and values
+	assertNoDataRace(t, &deepBidReqCopy, &shallowBidReqCopy, deepImpCopies, shallowImpCopies)
```

Comment:
The first thing that brought me to write `assertNoDataRace` was to have descriptive fail messages such as `"Data race in BidRequest.Site field"` whenever a race condition was found so it would be easier to locate the problem in the adapter source code. But now that we opted to copy with `deepcopy.Copy()`, that avoids the initialization of the zero-lenght slices, I believe assert.Equal()` simplifies the code and yields descriptive-enough test-fail messages